samcli/commands/check/resources/TemplateResource.py

```diff
@@ -0,0 +1,30 @@
+"""
+Super class for all resources
+"""
+from typing import Any
+
+
+class TemplateResource:
+    _resource_object: Any
+    _resource_type: str
+
+    def __init__(self, resource_object: Any, resource_type: str):
+        """
+        Args:
+            resource_object (Function): The resource object form the template file
+            resource_type (str): The resource type
+        """
+        self._resource_object = resource_object
+        self._resource_type = resource_type
+
+    @property
+    def resource_object(self) -> Any:
+        return self._resource_object
+
+    @property
+    def resource_name(self) -> Any:
+        return self._resource_object.name
```

Comment:
Since `_resource_object` is now `Any`, we don't know whether it has the attribute of `name` or not.
Getting the `name` from `Function` object should now be moved to `LambdaFunction` class.

```
class LambdaFunction(TemplateResource):
    ...
    @property
    def resource_name(self) -> str:
        return cast(Function, self._resource_object).name
```

If we still want to keep the contract that all `TemplateResource` instances should have `resource_name` property, then we can use `@abstractmethod`:

```
class TemplateResource:
     ...
    @property 
    @abstractmethod
    def resource_name(self) -> str:
        raise NotImplementedError()
```

Reply:
All `TemplateResources` should/will have this property. I'll make it an abstract method